File: client/mc/1.21.1-neoforge/src/main/java/calebxzhou/rdi/mc/client/rcmd/RcmdClientBridge211.java
package calebxzhou.rdi.mc.client.rcmd;

import calebxzhou.rdi.mc.common.RDI;
import calebxzhou.rdi.mc.common2.rcmd.client.RcmdClientBridge;
import calebxzhou.rdi.mc.common2.rcmd.client.RcmdLangSnapshot;
import calebxzhou.rdi.mc.common2.rcmd.client.RcmdRecipeSnapshot;
import calebxzhou.rdi.mc.common2.rcmd.client.RcmdRecipeView;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.language.ClientLanguage;
import net.minecraft.network.chat.Component;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public final class RcmdClientBridge211 implements RcmdClientBridge {
    private final Minecraft minecraft;

    public RcmdClientBridge211(Minecraft minecraft) {
        this.minecraft = minecraft;
    }

    @Override
    public String name() {
        return minecraft.getUser().getName();
    }

    @Override
    public UUID playerId() {
        return minecraft.player == null ? NO_PLAYER_ID : minecraft.player.getUUID();
    }

    @Override
    public boolean hasPermission(String permission) {
        return true;
    }

    @Override
    public void sendFeedback(String message) {
        sendMessage(message);
    }

    @Override
    public void sendError(String message) {
        sendMessage("[rcmd] " + message);
    }

    @Override
    public Path gameDirectory() {
        return minecraft.gameDirectory.toPath();
    }

    @Override
    public void executeOnMainThread(Runnable task) {
        minecraft.execute(task);
    }

    @Override
    public void setFirmChunkVisible(boolean visible) {
        RDI.SHOW_FIRM_CHUNKS = visible;
    }

    @Override
    public RcmdRecipeSnapshot recipeExportSnapshot() {
        if (minecraft.level == null) {
            return null;
        }
        var registries = minecraft.level.registryAccess();
        var recipes = new ArrayList<RcmdRecipeView>();
        for (var holder : minecraft.level.getRecipeManager().getOrderedRecipes()) {
            var recipe = holder.value();
            var result = recipe.getResultItem(registries);
            recipes.add(RcmdRecipeCodec211.recipeView(
                    holder.id().toString(),
                    recipe,
                    result
            ));
        }
        return new RcmdRecipeSnapshot(List.copyOf(recipes));
    }

    @Override
    public RcmdLangSnapshot langExportSnapshot() {
        var resourceManager = minecraft.getResourceManager();
        var english = ClientLanguage.loadFrom(resourceManager, List.of("en_us"), false).getLanguageData();
        var chinese = ClientLanguage.loadFrom(resourceManager, List.of("zh_cn"), false).getLanguageData();
        return new RcmdLangSnapshot(english, chinese);
    }

    private void sendMessage(String message) {
        var component = Component.literal(message);
        if (minecraft.player != null) {
            minecraft.player.displayClientMessage(component, false);
            return;
        }
        minecraft.gui.getChat().addMessage(component);
    }
}

File: client/mc/1.21.1-neoforge/src/main/kotlin/calebxzhou/rdi/mc/client/rcmd/RcmdClientBridge211.kt
package calebxzhou.rdi.mc.client.rcmd

import calebxzhou.rdi.mc.common.RDI
import calebxzhou.rdi.mc.common2.rcmd.client.RcmdClientBridge
import calebxzhou.rdi.mc.rcmd.RcmdSource
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
import java.nio.file.Path
import java.util.*

class RcmdClientBridge211(private val minecraft: Minecraft) : RcmdClientBridge {
    override fun name(): String {
        return minecraft.getUser().getName()
    }

    override fun playerId(): UUID {
        return if (minecraft.player == null) RcmdSource.NO_PLAYER_ID else minecraft.player!!.getUUID()
    }

    override fun hasPermission(permission: String?): Boolean {
        return true
    }

    override fun sendFeedback(message: String) {
        sendMessage(message)
    }

    override fun sendError(message: String?) {
        sendMessage("[rcmd] $message")
    }

    override fun gameDirectory(): Path {
        return minecraft.gameDirectory.toPath()
    }

    override fun executeOnMainThread(task: Runnable) {
        minecraft.execute(task)
    }

    override fun toggleSetFirmSectionsVisible(): Boolean {
        RDI.SHOW_SET_FIRM_SECTIONS = !RDI.SHOW_SET_FIRM_SECTIONS
        return RDI.SHOW_SET_FIRM_SECTIONS
    }

    override fun toggleNowFirmSectionVisible(): Boolean {
        RDI.SHOW_NOW_FIRM_SECTION = !RDI.SHOW_NOW_FIRM_SECTION
        return RDI.SHOW_NOW_FIRM_SECTION
    }


    private fun sendMessage(message: String) {
        val component = Component.literal(message)
        if (minecraft.player != null) {
            minecraft.player!!.displayClientMessage(component, false)
            return
        }
        minecraft.gui.getChat().addMessage(component)
    }
}
